sihl/src/core_container.ml

```diff
@@ -6,19 +6,35 @@ module Logs = (val Logs.src_log log_src : Logs.LOG)
 exception Exception
 
 type lifecycle =
-  { name : string
+  { type_name : string
+  ; implementation_name : string
+  ; id : int
   ; dependencies : unit -> lifecycle list
   ; start : unit -> unit Lwt.t
   ; stop : unit -> unit Lwt.t
   }
 
+let counter = ref 0
+
 let create_lifecycle
     ?(dependencies = fun () -> [])
     ?(start = fun () -> Lwt.return ())
     ?(stop = fun () -> Lwt.return ())
-    name
+    ?implementation_name
+    type_name
   =
-  { name; dependencies; start; stop }
+  (* Give all lifecycles unique names *)
+  counter := !counter + 1;
+  let implementation_name =
+    Option.value
+      implementation_name
+      ~default:(Format.asprintf "%s_impl" type_name)
```

Comment:
@aronerben I think you can leave `_impl` away since you are storing the `implementation_name`[

Reply:
Done